Memory leak inside circuits manager ?>

Hi,
after few tests,  it seems that circuits consume much more memory when using callEvent, than when using FireEvent with a waiting loop. To be more clear, I have written two script that process the same tasks, and theses scripts measure there memory consumption over iteration.

The two scripts can be found here: https://gist.github.com/y0no/09826e83e643bc0f98c2

One is using "call" and the other "fire". After execution here is the memory consumption for fire:
![Memory consumption on fire](http://i.imgur.com/UsBQm1H.png)

And this one for call:
![Memory consumption on call](http://i.imgur.com/5umc5nG.png)

As you can see call function use a lot more memory than fire.

In a perfect world, the memory consumption shoulld be the same in each case. And in a really perfect world the memory should not grow over time, because after each task the memory should be released.

After a few investigation, it seems that the issue come from manager who don't correctly stop iterator. It could be great to really stop iterator with a "StopIteration" exception that allow garbage collector to do its job.

launch 10 event is to small for measure the memory leak

test_memory_leak.py:

```Python
from circuits import Event, Component
import pytest
import resource


class test(Event):
    """test Event"""


class test_call(Event):
    """test Event"""


class App(Component):
    def test(self):
        return 'Hello World!'

    def test_call(self):
        for _ in range(10000):
            yield self.call(test())

    @property
    def memory(self):
        return resource.getrusage(resource.RUSAGE_SELF).ru_maxrss


@pytest.fixture
def app(request):
    app = App()

    def finalizer():
        app.stop()

    request.addfinalizer(finalizer)
    app.start()
    return app


def test_memory_call(app):
    start_memory = app.memory
    event = test_call()

    for _ in range(10000):
        for _ in app.call(event):
            pass

    assert app.memory == start_memory


def test_memory_fire(app):
    start_memory = app.memory

    for _ in range(10000):
        app.fire(test())

    while len(app):
        pass

    assert app.memory == start_memory


def test_memory_wait(app):
    start_memory = app.memory
    event = test()

    for _ in range(10000):
        for _ in app.waitEvent(event):
            pass

    assert app.memory == start_memory```

Test run:
```Shell
py.test tests/core/test_memory_leak.py -v
================================================================= test session starts =================================================================
platform linux -- Python 3.4.3 -- py-1.4.26 -- pytest-2.6.4 -- /usr/bin/python3
collected 3 items 

tests/core/test_memory_leak.py::test_memory_fire FAILED
tests/core/test_memory_leak.py::test_memory_wait FAILED
tests/core/test_memory_leak.py::test_memory_call FAILED

====================================================================== FAILURES =======================================================================
__________________________________________________________________ test_memory_fire ___________________________________________________________________

app = <App/* 10343:MainThread (queued=0) [R]>

    def test_memory_fire(app):
        start_memory = app.memory
    
        for _ in range(10000):
            app.fire(test())
    
        while len(app):
            pass
    
>       assert app.memory == start_memory
E       assert 31924 == 21832
E        +  where 31924 = <App/* 10343:MainThread (queued=0) [R]>.memory

tests/core/test_memory_leak.py:47: AssertionError
__________________________________________________________________ test_memory_wait ___________________________________________________________________

app = <App/* 10343:MainThread (queued=0) [R]>

    def test_memory_wait(app):
        start_memory = app.memory
        event = test()
    
        for _ in range(10000):
            for _ in app.waitEvent(event):
                pass
    
>       assert app.memory == start_memory
E       assert 50724 == 32148
E        +  where 50724 = <App/* 10343:MainThread (queued=0) [R]>.memory

tests/core/test_memory_leak.py:58: AssertionError
__________________________________________________________________ test_memory_call ___________________________________________________________________

app = <App/* 10343:MainThread (queued=7830) [R]>

    def test_memory_call(app):
        start_memory = app.memory
        event = test_call()
    
        for _ in range(10000):
            for _ in app.call(event):
                pass
    
>       assert app.memory == start_memory
E       assert 83608 == 50732
E        +  where 83608 = <App/* 10343:MainThread (queued=7831) [R]>.memory

tests/core/test_memory_leak.py:69: AssertionError
============================================================== 3 failed in 4.77 seconds ===============================================================
```

--- a/circuits/core/manager.py
+++ b/circuits/core/manager.py
@@ -544,8 +544,7 @@ class Manager(object):
 
         yield state
 
-        if not state.timeout:
-            self.removeHandler(_on_done_handler, "%s_done" % event_name)
+        self.removeHandler(_on_done_handler, "%s_done" % event_name)
 
         if state.event is not None:
             yield CallValue(state.event.value)

--- a/tests/core/test_memory_leaks.py
+++ b/tests/core/test_memory_leaks.py
@@ -1,0 +1,46 @@
+#!/usr/bin/env python
+import pytest
+
+from circuits import handler, Component, Event
+
+
+class call(Event):
+
+    """call Event"""
+    success = True
+
+
+class hello(Event):
+
+    """hello Event"""
+    success = True
+
+
+class App(Component):
+    @handler("call")
+    def _on_call(self):
+        x = yield self.call(hello())
+        yield x.value
+
+    def hello(self):
+        return "Hello World!"
+
+
+@pytest.fixture
+def app(request, manager, watcher):
+    app = App().register(manager)
+    assert watcher.wait("registered")
+
+    def finalizer():
+        app.unregister()
+
+    request.addfinalizer(finalizer)
+
+    return app
+
+
+def test_done_handlers_dont_leak(manager, watcher, app):
+    manager.fire(call())
+    manager.fire(call())
+    assert watcher.wait("call_success")
+    assert "hello_done" not in app._handlers

